### PR TITLE
draft: Fix issue with accessing null attribues for active record objects

### PR DESCRIPTION
**Pull request by trigger202**
Do not review yet, working on unit tests

### CODE REVIEW

File: src/Extension/CoreExtension.php [hunk 1]
```diff
@@ -1478,7 +1478,13 @@ function twig_get_attribute(Environment $env, Source $source, $object, $item, ar
 
     // object property
     if (/* Template::METHOD_CALL */ 'method' !== $type) {
-        if (isset($object->$item) || \array_key_exists((string) $item, (array) $object)) {
+//        if (isset($object->$item) || \array_key_exists((string) $item, (array) $object)) {
+        /**
+         * NOTE: this change is added fix issue when ActiveRecord object has the property but its value is null. e.g booking.agentId.
```
Comment (stof) [inline src/Extension/CoreExtension.php]: A comment saying `this change` is totally useless, as the comment is not describing a change (commit messages do that) but the code in the class.

And keeping the old line as a comment is also totally useless (keeping the history is handled by git in a much better way)

**Review by stof — CHANGES_REQUESTED** [review]
This should have tests to prevent  regressions

File: src/Extension/CoreExtension.php (hunk 1 above)
Comment (trigger202) [reply src/Extension/CoreExtension.php]: Thanks @stof for the feedback, I have removed it.

File: src/Extension/CoreExtension.php [hunk 2]
```diff
@@ -1655,4 +1655,29 @@ function twig_check_arrow_in_sandbox(Environment $env, $arrow, $thing, $type)
...
+
+    /**
+     * Check if object attribute regardless of the value, it should return true
+     * It should be able to handle when object is ActiveRecord and have null value for a column e.g $user.RoleId
+     *
+     * @param $object
+     * @param $attribute
+     * @return bool
+     */
+    function hasAttribute($object, $attribute)
```
Comment (fabpot) [inline src/Extension/CoreExtension.php:1667]:
```suggestion
    private function hasAttribute($object, $attribute)
```

File: src/Extension/CoreExtension.php [hunk 3]
```diff
@@ -1655,4 +1655,29 @@ function twig_check_arrow_in_sandbox(Environment $env, $arrow, $thing, $type)
...
+    {
+        if (!is_object($object)) {
```
Comment (fabpot) [inline src/Extension/CoreExtension.php]: This check can be removed as this is always an object at this point (see line 1455).

File: src/Extension/CoreExtension.php [hunk 4]
```diff
@@ -1655,4 +1655,29 @@ function twig_check_arrow_in_sandbox(Environment $env, $arrow, $thing, $type)
...
+            return false;
+        }
+
+        try {
+            $object->$attribute;
+
+            return true;
+        } catch (Exception $exception) {
```
Comment (fabpot) [inline src/Extension/CoreExtension.php]: Does PHP throw an exception when a property does not exist? That would be `\Exception` anyway, but I think it just emits a warning, right?

**Review by fabpot — CHANGES_REQUESTED** [review]
Without extensive tests, it's not something we will merge.

File: src/Extension/CoreExtension.php (hunk 4 above)
Comment (trigger202) [reply src/Extension/CoreExtension.php]: I believe you're right, accessing random attributes didn't trigger an exception, thanks.

File: src/Extension/CoreExtension.php [hunk 5]
```diff
@@ -1666,18 +1666,10 @@ function twig_check_arrow_in_sandbox(Environment $env, $arrow, $thing, $type)
...
-        }
-
-        try {
-            $object->$attribute;
-
-            return true;
-        } catch (Exception $exception) {
-            // do nothing
+        if (method_exists($object, 'hasAttribute')) {
+            return $object->hasAttribute($attribute);
```
Comment (smnandre) [inline src/Extension/CoreExtension.php:1674]: This seems not something generic or universal ... or is this method part of some standard ?

File: src/Extension/CoreExtension.php (hunk 2 above)
Comment (trigger202) [reply src/Extension/CoreExtension.php:1667]: making the method private is making all the exists tests fail
image

**Comment by trigger202** [comment]
> Without extensive tests, it's not something we will merge.

Okay, just wondering how to turn this PR into a draft so I can ask for a code review once I feel its ready. if someone can point out how to make it a draft

File: src/Extension/CoreExtension.php (hunk 5 above)
Comment (trigger202) [reply src/Extension/CoreExtension.php:1674]: I think most active record patterns have this method. Yii and Laravel do at least couldnt find anything for others

https://github.com/illuminate/database/blob/master/Eloquent/Concerns/HasAttributes.php#L45

Comment (fabpot) [reply src/Extension/CoreExtension.php:1674]: That's not something we will add to Twig core.

**Pull request closed without merging**